### PR TITLE
Support TaskExt in CliApp

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**
Misses release notes and tests but can already be tried out by early adopters ;)

### CODE REVIEW

File: src/ocean/util/app/CliApp.d [hunk 1]
```diff
@@ -97,6 +107,24 @@ abstract class CliApp : Application, IArgumentsExtExtension
...
+            applications already configuring scheduler on their own.
+
+        ***********************************************************************/
+
+        bool use_task_ext;
+
+        /***********************************************************************
+
+            Only used if `use_task_ext` is set to `true`. Defines scheduler
+            configuratio to be used by TaskExt.
```
Comment (joseph-wakeling-sociomantic) [inline src/ocean/util/app/CliApp.d]: Spelling: `configuration` ;-)

**Comment by joseph-wakeling-sociomantic** [comment]
I'm trying this with a small CLI app and getting a runtime failure:

```
assertion failed at ./submodules/ocean/src/ocean/task/Scheduler.d:865
  toString():  Scheduler is null, initScheduler must be called before using it
Aborted (core dumped)
```

App constructor:

```D
    public this ()
    {
        typeof(this).OptionalSettings settings;
        settings.use_task_ext = true;
        super(this.app_name, this.app_desc, version_info, settings);
    }
```

App `run` method:

```D
    public override int run (Arguments args)
    {
        verify(Task.getThis() !is null);
        return 0;
    }
```

**Comment by mihails-strasuns-sociomantic** [comment]
Updated:
- test
- release notes
- fixed implementation :)

**Comment by joseph-wakeling-sociomantic** [comment]
Yea, this works with my app.  I'll re-review in detail once the couple of issues identified by CI are fixed.

**Comment by mihails-strasuns-sociomantic** [comment]
Updated:
- fix 2.071 build

File: integrationtest/taskext_cli/main.d [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+*******************************************************************************/
+
+module integrationtest.taskext_cli.main;
+
+import ocean.transition;
+import ocean.util.app.CliApp;
+import ocean.text.Arguments;
+import ocean.task.Task;
+import ocean.task.Scheduler;
+import ocean.core.Test;
```
Comment (joseph-wakeling-sociomantic) [inline integrationtest/taskext_cli/main.d:23]: Minor style: imports can/should be in alphabetical order?

File: src/ocean/util/app/CliApp.d [hunk 3]
```diff
@@ -24,6 +24,7 @@ module ocean.util.app.CliApp;
 
 import ocean.util.app.Application : Application;
 import ocean.util.app.ext.model.IArgumentsExtExtension;
+import ocean.task.IScheduler;
```
Comment (joseph-wakeling-sociomantic) [inline src/ocean/util/app/CliApp.d:27]: Alphabetical order of imports?

File: src/ocean/util/app/CliApp.d [hunk 4]
```diff
@@ -38,6 +39,7 @@ abstract class CliApp : Application, IArgumentsExtExtension
     import ocean.text.Arguments : Arguments;
     import ocean.util.app.ext.ArgumentsExt;
     import ocean.util.app.ext.VersionArgsExt;
+    import ocean.util.app.ext.TaskExt;
```
Comment (joseph-wakeling-sociomantic) [inline src/ocean/util/app/CliApp.d:42]: Alphabetical order of imports?

**Review by joseph-wakeling-sociomantic — APPROVED** [review]
Bar a couple of minor style issues, LGTM.

File: integrationtest/taskext_cli/main.d (hunk 2 above)
Comment (mihails-strasuns-sociomantic) [reply integrationtest/taskext_cli/main.d:23]: Nope, we don't require that. As a rule for automatic formatter that can be an option, but it is not worth bothering for manual effort.

File: src/ocean/util/app/ext/TaskExt.d [hunk 5]
```diff
@@ -44,6 +44,30 @@ import ocean.meta.codegen.Identifier;
...
 class TaskExt : IConfigExtExtension
 {
+    /***************************************************************************
+
+        Default scheduler configuration to use in absence of explicit
+        configuration setting.
+
+    ***************************************************************************/
+
+    SchedulerConfiguration config;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/app/ext/TaskExt.d:54]: Why does `DaemonApp` not allow this to be set?

Comment (mihails-strasuns-sociomantic) [reply src/ocean/util/app/ext/TaskExt.d:54]: No particular reason - it just wasn't of much use when configuration file is present. I will add it for consistency.

Comment (mihails-strasuns-sociomantic) [reply src/ocean/util/app/ext/TaskExt.d:54]: Updated

File: relnotes/taskextcli.feature.md [hunk 6]
```diff
@@ -25,3 +25,6 @@ class TestApp : CliApp
     }
 }
 ```
+
```
Comment (gavin-norman-sociomantic) [inline relnotes/taskextcli.feature.md]: I'd put this as a separate note, otherwise `DaemonApp` users will miss it under the heading `CliApp now supports TaskExt`.

File: src/ocean/util/app/ext/TaskExt.d [hunk 7]
```diff
@@ -44,6 +44,30 @@ import ocean.meta.codegen.Identifier;
 /// ditto
 class TaskExt : IConfigExtExtension
 {
+    /***************************************************************************
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/app/ext/TaskExt.d:47]: Maybe make this a separate commit?

File: src/ocean/util/app/ext/TaskExt.d (hunk 5 above)
Comment (gavin-norman-sociomantic) [reply src/ocean/util/app/ext/TaskExt.d:54]: :+1: yeah, that makes more sense. I wasn't sure whether there was some subtlety that I was missing.

**Comment by mihails-strasuns-sociomantic** [comment]
Updated

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Comment by gavin-norman-sociomantic** [comment]
Needs rebasing, apparently.

**Comment by mihails-strasuns-sociomantic** [comment]
@gavin-norman-sociomantic I have suspicion github has updated UI to ask for rebasing even if there are no conflicts - only PRs that are based exactly on HEAD become green. You can still force rebase/merge it as administrator.

**Comment by joseph-wakeling-sociomantic** [comment]
Yea, it's a safety mechanism that's intended to ensure CI is run for the PR features together with the very latest changes in the target branch.  FWIW we could probably guarantee the same result by making sure that we test merges into target branch as well as the PR alone (but of course GitHub can't tell if we are doing that, so it sounds this note of caution).
